tools/tune/eq/gui.m

```diff
@@ -0,0 +1,476 @@
+  EQ_right = uipanel ("title", "EQ Right", "position", [0, 0, 1, 0.5]);
+  EQ_left = uipanel ("title", "EQ Left", "position", [0, 0.5, 1, 0.5]);
+  main_panes = [struct("pane", EQ_left); struct("pane", EQ_right)];
+
+  % UI panel relative position table
+  positions = [
+    [0    0   0.25 0.5 ]
+    [0.25 0   0.5  0.5 ]
+    [0.5  0   0.75 0.5 ]
+    [0.75 0   1    0.5 ]
+    [0    0.5 0.25 1   ]
+    [0.25 0.5 0.5  1   ]
+    [0.5  0.5 0.75 1   ]
+    [0.75 0.5 1    1   ]
+  ];
+
+  fn = figure(fn); fn = fn + 1;
+  h = [0 0];
+  w = [1 nyquist_f];
+  n_octaves = 10;
+  subplot(2, 1, 1);
+  main_panes(1).plot = plot(log2(w), h, "linewidth", 4);
+  % Range set to match http://audio-tuning.appspot.com/
+  axis([log2(nyquist_f / (2 ^ n_octaves)), log2(nyquist_f), -24, 18]);
+  subplot(2, 1, 2);
+  main_panes(2).plot = plot(log2(w), h, "linewidth", 4);
+  % Range set to match http://audio-tuning.appspot.com/
+  axis([log2(nyquist_f / (2 ^ n_octaves)), log2(nyquist_f), -24, 18]);
+  for c = 1:2
+    for i = 1:length(positions)
+      EQ = initialize_biquad(i, c, positions(i, 1:end), main_panes, Fs, EQ);
+    endfor
+  endfor
+  % octave is silly and doesn't support references so we have to re-write the
+  % control matrix so the callbacks have access without using globals
+  for c = 1:2
+    for i = 1:length(positions)
+      update_control_elements(EQ(i, c), EQ);
+    endfor
+  endfor
+  fn = figure(fn, "toolbar", "none", "menubar", "none", 'Position', [200 200 300 190]); fn = fn + 1;
+  menu_window = struct();
+  menu_window.ip_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 160 180 30]);
+  menu_window.ip_label = uicontrol( ...
+        "style", "text",
+        "string", "DUT IP:", ...
+        "position", [0 160 120 30]);
+  menu_window.device_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 130 180 30]);
+  menu_window.device_label = uicontrol( ...
+        "style", "text",
+        "string", "Device Num:", ...
+        "position", [0 130 120 30]);
+  menu_window.alsa_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 100 180 30]);
+  menu_window.alsa_label = uicontrol( ...
+        "style", "text",
+        "string", "ALSA Ctl Num:", ...
+        "position", [0 100 120 30]);
+  menu_window.upload = uicontrol( ...
+        "style", "pushbutton", ...
+        'string', 'Push config to DUT', ...
+        "callback", @push_config, ...
+        "position", [0 50 300 50], ...
+        "userdata", struct("menu_window", menu_window, "EQ", EQ));
+  menu_window.load_dsp = uicontrol( ...
+        "style", "pushbutton", ...
+        'string', 'Load dsp.ini', ...
+        "callback", @eq_load_dsp_ini, ...
+        "position", [0 0 300 50], ...
+        "userdata", struct("EQ", EQ));
+end
+
+function conf = get_config()
+  % since octave has no definitive way to constants we are going to hack a
+  % struct into behaving like one
+  conf = struct( ...
+      "Fs", 48000 ...
```

Comment:
It seems like push_config() gets the sample rate from eq_defaults(). We should avoid duplication and move preferences to one location. What would you prefer?

Reply:
agreed, these should be unified. I will get back to you on which we should use, I am currently thinking that we should use the GUI's value and push it into the EQ config.

Reply:
Yep, it's best to get the value from GUI for all the filter design functions.